File: CFWebControls/GroupedDropdownList.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web.UI.WebControls;

namespace CFWebControls
{
    public class GroupedDropdownList : DropDownList 
    {
        public const string OptionGroupTag = "optgroup";
        public const string OptionTag = "option";
        protected override void RenderContents(System.Web.UI.HtmlTextWriter writer)
        {
            ListItemCollection items = this.Items;
            int count = items.Count;
            string tag;
            string optgroupLabel;
            if (count > 0)
            {
                bool flag = false;
                string prevOptGroup = null;
                for (int i = 0; i <= count - 1; i++)
                {
                    tag = OptionTag;
                    optgroupLabel = null;
                    ListItem item = items[i];
                    if (item.Enabled)
                    {
                        if (item.Attributes != null && item.Attributes.Count > 0 && item.Attributes[OptionGroupTag] != null)
                        {
                            optgroupLabel = item.Attributes[OptionGroupTag];
                            if (prevOptGroup != optgroupLabel)
                            {
                                if (prevOptGroup != null)
                                {
                                    writer.WriteEndTag(OptionGroupTag);
                                }

                                writer.WriteBeginTag(OptionGroupTag);
                                if (!string.IsNullOrEmpty(optgroupLabel))
                                {
                                    writer.WriteAttribute("label", optgroupLabel);
                                }

                                writer.Write('>');
                            }

                            item.Attributes.Remove(OptionGroupTag);
                            prevOptGroup = optgroupLabel;
                        }
                        else
                        {
                            if (prevOptGroup != null)
                            {
                                writer.WriteEndTag(OptionGroupTag);
                            }

                            prevOptGroup = null;
                        }

                        writer.WriteBeginTag(tag);
                        if (item.Selected)
                        {
                            if (flag)
                            {
                                this.VerifyMultiSelect();
                            }

                            flag = true;
                            writer.WriteAttribute("selected", "selected");
                        }

                        writer.WriteAttribute("value", item.Value, true);
                        if (item.Attributes != null && item.Attributes.Count > 0)
                        {
                            item.Attributes.Render(writer);
                        }

                        if (optgroupLabel != null)
                        {
                            item.Attributes.Add(OptionGroupTag, optgroupLabel);
                        }

                        if (this.Page != null)
                        {
                            this.Page.ClientScript.RegisterForEventValidation(this.UniqueID, item.Value);
                        }

                        writer.Write('>');
                        System.Web.HttpUtility.HtmlEncode(item.Text, writer);
                        writer.WriteEndTag(tag);
                        writer.WriteLine();
                        if (i == count - 1)
                        {
                            if (prevOptGroup != null)
                            {
                                writer.WriteEndTag(OptionGroupTag);
                            }
                        }
                    }
                }
            }
        }

        protected override object SaveViewState()
        {
            object[] state = new object[this.Items.Count + 1];
            object baseState = base.SaveViewState();
            state[0] = baseState;
            bool itemHasAttributes = false;
            for (int i = 0; i <= this.Items.Count - 1; i++)
            {
                if (this.Items[i].Attributes.Count > 0)
                {
                    itemHasAttributes = true;
                    object[] attributes = new object[this.Items[i].Attributes.Count * 2 - 1 + 1];
                    int k = 0;
                    foreach (string key in this.Items[i].Attributes.Keys)
                    {
                        attributes[k] = key;
                        k += 1;
                        attributes[k] = this.Items[i].Attributes[key];
                        k += 1;
                    }

                    state[i + 1] = attributes;
                }
            }

            if (itemHasAttributes)
            {
                return state;
            }

            return baseState;
        }

        protected override void LoadViewState(object savedState)
        {
            if (null == savedState) return;
            var state = savedState as ArrayList;
            if (savedState.GetType().GetElementType() != null &&
                savedState.GetType().GetElementType().Equals(typeof(Object)))
            {
                
                base.LoadViewState(state[0]);
            
                for(int i = 1; i < state.Count-1; i++)
                {
                    if (state[1] != null)
                    {
                        var attributes = (ArrayList)state[i];
                        for (int k = 0; k < attributes.Count - 1; k++)
                        {
                            this.Items[i - 1].Attributes.Add(attributes[k].ToString(), attributes[k + 1].ToString());
                        }
                    }
                }
            }
            else
            {
                base.LoadViewState(savedState);
            }
        }
    }
}

File: cleanfleets-fms/controls/GroupedDropDownList.vb


<Assembly: TagPrefix("cleanfleets", "CF")>


Public Class GroupedDropDownList
	Inherits System.Web.UI.WebControls.DropDownList
	Public Const OptionGroupTag As String = "optgroup"
	Private Const OptionTag As String = "option"
	Protected Overrides Sub RenderContents(writer As System.Web.UI.HtmlTextWriter)
		Dim items As ListItemCollection = Me.Items
		Dim count As Integer = items.Count
		Dim tag As String
		Dim optgroupLabel As String
		If count > 0 Then
			Dim flag As Boolean = False
			Dim prevOptGroup As String = Nothing
			For i As Integer = 0 To count - 1
				tag = OptionTag
				optgroupLabel = Nothing
				Dim item As ListItem = items(i)
				If item.Enabled Then
					If item.Attributes IsNot Nothing AndAlso item.Attributes.Count > 0 AndAlso item.Attributes(OptionGroupTag) IsNot Nothing Then
						optgroupLabel = item.Attributes(OptionGroupTag)

						If prevOptGroup <> optgroupLabel Then
							If prevOptGroup IsNot Nothing Then
								writer.WriteEndTag(OptionGroupTag)
							End If
							writer.WriteBeginTag(OptionGroupTag)
							If Not String.IsNullOrEmpty(optgroupLabel) Then
								writer.WriteAttribute("label", optgroupLabel)
							End If
							writer.Write(">"C)
						End If
						item.Attributes.Remove(OptionGroupTag)
						prevOptGroup = optgroupLabel
					Else
						If prevOptGroup IsNot Nothing Then
							writer.WriteEndTag(OptionGroupTag)
						End If
						prevOptGroup = Nothing
					End If

					writer.WriteBeginTag(tag)
					If item.Selected Then
						If flag Then
							Me.VerifyMultiSelect()
						End If
						flag = True
						writer.WriteAttribute("selected", "selected")
					End If
					writer.WriteAttribute("value", item.Value, True)
					If item.Attributes IsNot Nothing AndAlso item.Attributes.Count > 0 Then
						item.Attributes.Render(writer)
					End If
					If optgroupLabel IsNot Nothing Then
						item.Attributes.Add(OptionGroupTag, optgroupLabel)
					End If
					If Me.Page IsNot Nothing Then
						Me.Page.ClientScript.RegisterForEventValidation(Me.UniqueID, item.Value)
					End If

					writer.Write(">"C)
					HttpUtility.HtmlEncode(item.Text, writer)
					writer.WriteEndTag(tag)
					writer.WriteLine()
					If i = count - 1 Then
						If prevOptGroup IsNot Nothing Then
							writer.WriteEndTag(OptionGroupTag)
						End If
					End If
				End If
			Next
		End If
	End Sub

	Protected Overrides Function SaveViewState() As Object
		Dim state As Object() = New Object(Me.Items.Count) {}
		Dim baseState As Object = MyBase.SaveViewState()
		state(0) = baseState
		Dim itemHasAttributes As Boolean = False

		For i As Integer = 0 To Me.Items.Count - 1
			If Me.Items(i).Attributes.Count > 0 Then
				itemHasAttributes = True
				Dim attributes As Object() = New Object(Me.Items(i).Attributes.Count * 2 - 1) {}
				Dim k As Integer = 0

				For Each key As String In Me.Items(i).Attributes.Keys
					attributes(k) = key
					k += 1
					attributes(k) = Me.Items(i).Attributes(key)
					k += 1
				Next
				state(i + 1) = attributes
			End If
		Next

		If itemHasAttributes Then
			Return state
		End If
		Return baseState
	End Function

	Protected Overrides Sub LoadViewState(savedState As Object)
		If savedState Is Nothing Then
			Return
		End If

		If Not (savedState.[GetType]().GetElementType() Is Nothing) AndAlso (savedState.[GetType]().GetElementType().Equals(GetType(Object))) Then
			Dim state As Object() = DirectCast(savedState, Object())
			MyBase.LoadViewState(state(0))

			For i As Integer = 1 To state.Length - 1
				If state(i) IsNot Nothing Then
					Dim attributes As Object() = DirectCast(state(i), Object())
					For k As Integer = 0 To attributes.Length - 1 Step 2
						Me.Items(i - 1).Attributes.Add(attributes(k).ToString(), attributes(k + 1).ToString())
					Next
				End If
			Next
		Else
			MyBase.LoadViewState(savedState)
		End If
	End Sub
End Class


'=======================================================
'Service provided by Telerik (www.telerik.com)
'Conversion powered by NRefactory.
'Twitter: @telerik
'Facebook: facebook.com/telerik
'=======================================================

' (http://stackoverflow.com/questions/130020/dropdownlist-control-with-optgroups-for-asp-net-webforms?lq=1)


